Move LoadError logic into base Providers

HaoK opened the pull request:
@divega @glennc 

This has a breaking change component to it, in that we want to take over the default implementation of IConfigurationProvider Load and make that non virtual (and have derrived classes implement DoLoad instead), otherwise it'd be easy to lose the LoadExceptionHandling logic in base...

The good news is it's a compilation error for any existing overrides of Load()...

Thoughts?

HaoK commented:
Fixes https://github.com/aspnet/Configuration/issues/572

divega commented on src/Microsoft.Extensions.Configuration.Abstractions/LoadExceptionContext.cs:
```diff
@@ -6,14 +6,14 @@
+    /// Contains information about a load exception.
     /// </summary>
-    public class FileLoadExceptionContext
+    public class LoadExceptionContext
```
I think it could be a good idea to give this more of a unique name independent of the namespace, e.g. `ConfigurationLoadExceptionContext`

divega commented on src/Microsoft.Extensions.Configuration.FileExtensions/FileConfigurationSource.cs:
```diff
@@ -39,9 +39,9 @@ public abstract class FileConfigurationSource : IConfigurationSource
+        /// Will be called if an uncaught exception occurs in Load.
         /// </summary>
-        public Action<FileLoadExceptionContext> OnLoadException { get; set; }
+        public Action<LoadExceptionContext> OnLoadException { get; set; }
```
We could lift this property to the `IConfigurationSource` interface. Thoughts?

divega replied:
It would probably imply making this more of first class concept, even on IConfiguraitonProvider.

HaoK replied:
Yeah we can make it a first class required concept I guess

divega replied:
Ok, BTW, I am not convinced one way or the other :smile:

divega commented on src/Microsoft.Extensions.Configuration/ConfigurationProvider.cs:
```diff
@@ -53,10 +58,40 @@ public virtual void Set(string key, string value)
+        /// <summary>
+        /// Loads (or reloads) the data for this provider.
+        /// </summary>
+        public virtual void DoLoad()
```
protected?

divega replied:
abstract?

HaoK replied:
Today Load isn't abstract since most of our implementations actually do nothing.  So a no-op DoLoad was parity, my first feeling was that it should be abstract too :)

ycrumeyrolle replied:
Is the method DoLoad name the most adapted name?
DoLoad, LoadCore, LoadImpl, LoadConfiguration, ...?

ycrumeyrolle replied:
Is the method DoLoad name the most adapted name?
DoLoad, LoadCore, LoadImpl, LoadConfiguration, ...?

divega replied:
@ycrumeyrolle No, I don't think `DoLoad` is a great name, but I was going to wait until @HaoK sends an update to see where things land.

In the meanwhile, I would add `LoadData`  to the candidate list.

HaoK replied:
So if we are going to promote the delegate to IConfigurationSource, we can actually do this cleaner now via a `LoadSafe` extension method to IConfigurationProvider, we'll need to add the Source property to IConfigurationProvider, but with that we'll be able to move this logic into an extension method instead of doing naming tricks/wrapping a virtual Load.  Breaking changes on the interfaces instead of mucking with Load :)

divega replied:
I see. I was starting converge to this too. I think we should have a general conversation about breaking changes soon :smile:

divega commented on src/Microsoft.Extensions.Configuration.FileExtensions/FileConfigurationProvider.cs:
```diff
@@ -26,6 +26,7 @@ public FileConfigurationProvider(FileConfigurationSource source)
                 throw new ArgumentNullException(nameof(source));
             }
             Source = source;
+            OnLoadException = source.OnLoadException;
```
Make `OnLoadException` property virtual and have the one on `FileConfigurationProvider` access `source.OnLoadException`?

divega commented on src/Microsoft.Extensions.Configuration.FileExtensions/FileConfigurationProvider.cs:
```diff
@@ -104,7 +84,7 @@ private void Load(bool reload)
         /// <exception cref="FileNotFoundException">If Optional is <c>false</c> on the source and a
         /// file does not exist at specified Path.</exception>
-        public override void Load()
+        public override void DoLoad()
```
Curious, isn't the call to `Load(reload: true) that happens in the delegate we pass to `ChangeToken.OnChange()` also supposed to handle erros? I might be a bit confused already but from looking at the code, it seems to me that it doesn't because it doesn't go through the parameter-less `Load()`.

HaoK replied:
Yep good catch, looks like a missing test too, I'll add a test :)

divega replied:
BTW, should the version of the method that takes a stream in the file base provider also handle errors?

HaoK replied:
Well the ideal place we end up here is that providers should no longer need to care about errors, the error handling happens outside of provider code... so the answer should eventually be no before we merge :)

divega replied:
> Well the ideal place we end up here is that providers should no longer need to care about errors, the error handling happens outside of provider code... 

Ok. Should we move the error handling to a utility method we use any time we invoke Load? Or template method pattern is enough?

HaoK commented:
Updated, this has turned into a pretty big change...

- Added two extension methods SafeLoad/SafeInvoke(needed for File reloads to safely invoke Load(stream))
- Added ConfigurationSource abstract base to contain the new OnLoadException property, also standardized the pattern for all Providers to have a ctor/property that takes the source (Azure/Command/Env were non standard).
- Added a test verifying that reload errors were actually caught as well.

HaoK commented:
Oh, yeah the big change was making ConfigurationProvider a generic base class now to get TSource to be typed property (yuck...but seems worth it)

ycrumeyrolle commented on src/Microsoft.Extensions.Configuration.Abstractions/ConfigurationExtensions.cs:
```diff
@@ -68,5 +69,82 @@ public static bool Exists(this IConfigurationSection section)
+                }
+                if (!ignoreException)
+                {
+                    throw e;
```
`throw;` instead of `throw e;` ?

HaoK replied:
Yeah good catch

HaoK replied:
InvokeWithExceptionHandler/LoadWithExceptionHandler

HaoK commented:
Not doing for now
